source/common/router/scoped_rds.h

```diff
@@ -183,6 +189,38 @@ class ScopedRoutesConfigProviderManager : public Envoy::Config::ConfigProviderMa
       std::vector<std::unique_ptr<const Protobuf::Message>>&& config_protos,
       Server::Configuration::FactoryContext& factory_context,
       const Envoy::Config::ConfigProviderManager::OptionalArg& optarg) override;
+
+  RouteConfigProviderManager& route_config_provider_manager() {
+    return route_config_provider_manager_;
+  }
+
+  std::shared_ptr<RouteConfigProvider>
+  getRouteConfigProvider(Server::Configuration::FactoryContext& factory_context,
+                         envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
+                         std::string stat_name) {
+    const uint64_t route_provider_id = MessageUtil::hash(rds);
+    auto iter = cached_route_providers_.find(route_provider_id);
+    if (iter != cached_route_providers_.end()) {
+      return iter->second;
+    }
+    std::shared_ptr<RouteConfigProvider> provider =
+        route_config_provider_manager_.createRdsRouteConfigProvider(rds, factory_context,
+                                                                    stat_name);
+    cached_route_providers_.try_emplace(route_provider_id, provider);
+    return provider;
+  }
+  void
+  deleteRouteConfigSource(envoy::config::filter::network::http_connection_manager::v2::Rds& rds) {
+    const uint64_t route_provider_id = MessageUtil::hash(rds);
+    if (!cached_route_providers_.erase(route_provider_id)) {
+      ENVOY_LOG_MISC(warn, "Try to delete non-existing route config source {}", rds.DebugString());
+    }
+  }
+
+private:
+  RouteConfigProviderManager& route_config_provider_manager_;
+  // From ConfigSource fingerprint to shared RouteConfigProvider.
+  absl::flat_hash_map<uint64_t, std::shared_ptr<RouteConfigProvider>> cached_route_providers_;
```

Comment:
This is an object lifetime management simplification as opposed to an optimization. My concern is that the current approach requires the cache to be explicitly cleared when a RouteConfigProvider is no longer used (i.e., the ScopedRouteInfo is destroyed). Using a std::weak_ptr allows you to get rid of deleteRouteConfigSource() since this class won't actually store the shared_ptr.

Reply:
holding all the stale scope keys in the map doesn't look right to me. I'd rather we remove/delete explicitly when the routeConfigProvider is not used any more. As long as we are doing the right thing, deleting the cached provider should not be a problem.